flink-tests/src/test/java/org/apache/flink/test/scheduling/ReactiveModeITCase.java

```diff
@@ -196,52 +192,91 @@ public void run(SourceContext<String> ctx) throws Exception {
         public void cancel() {
             running = false;
         }
+
+        @Override
+        public void close() throws Exception {
+            tracker.reportStoppedInstance();
+        }
     }
 
-    private static class ParallelismTrackingSink<T> extends RichSinkFunction<T> {
+    private static class InstanceParallelismTracker {
+        // only notify this lock on scale-up
+        private final Object lock = new Object();
 
-        private static final InstanceTracker instances = new InstanceTracker();
+        private volatile int instances = 0;
```

Comment:
Why is this field volatile?

Reply:
I wasn't aware that synchronization also creates a "happens-before" memory barriers. Thanks a lot!